Kuadrant/multicluster-gateway-controller
# Gh 583

**What**

makes some significant changes to placement logic
adds finalizers to placement decisions
ensure gateways are removed when placement decisions are deleted 
ensure changing placement choice cleans up existing gateways
adds a placement condition to the gateway status

- [x] Update integration tests
- [x] Update e2e tests


**Verification**

- use the quickstart
- scale down the mgc controller
- run this version locally

Create a Gateway
```
apiVersion: gateway.networking.k8s.io/v1beta1
kind: Gateway
metadata:
  name: prod-web
  namespace: multi-cluster-gateways
  labels:
    cluster.open-cluster-management.io/placement: http-gateway
spec:
  gatewayClassName: kuadrant-multi-cluster-gateway-instance-per-cluster
  listeners:
  - allowedRoutes:
      namespaces:
        from: All
    name: specific
    hostname: 'specific.cb.hcpapps.net'
    port: 443
    protocol: HTTP
```

This will place it on the CP cluster
Verify the gateway is placed (should be an istio gateway present)
Verify the status is updated as expected on the gateway

```
 - lastTransitionTime: "2023-09-28T11:06:30Z"
    message: http-gateway
    observedGeneration: 1
    reason: PlacementCompleted
    status: "True"
    type: Placed
```

Create a new placement

```
apiVersion: cluster.open-cluster-management.io/v1beta1
kind: Placement
metadata:
  name: gateways
  namespace: multi-cluster-gateways
spec:
  numberOfClusters: 2
  clusterSets:
    - gateway-clusters
```

make sure second cluster availabe for placement
```
k label managedcluster kind-mgc-workload-1 ingress-cluster=true
```

update the placement label on the gateway to use this

```
k label gateway prod-web -n multi-cluster-gateways "cluster.open-cluster-management.io/placement"=gateways --overwrite
```

check the gateway is now on 2 clusters and placement condition is set to true

delete the placement

```
k delete placement gateways -n multi-cluster-gateways

```

check istio gateways are all gone. Check status is set to placement not found

Update the placement label

```

k label gateway prod-web -n multi-cluster-gateways "cluster.open-cluster-management.io/placement"=http-gateway --overwrite
```

check it is just on one cluster and the status is correct
set it back to something that doesn't exist
```

k label gateway prod-web -n multi-cluster-gateways "cluster.open-cluster-management.io/placement"=nope --overwrite
```

delete the gateway

```
k delete gateway prod-web -n multi-cluster-gateways
```
the gateway should delete
